Chapter three, first draft for review. (#11)
riscv-non-isa/riscv-security-model: specification/src/chapter3.adoc

ePMP extends PMP protection by allowing machine mode to restrict its own access to memory allocated to lower privilege levels. This can be used to mitigate against privilege escalation attacks, for example.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  
| If PMP is supported then ePMP MUST be supported.
|===

==== sPMP

https://github.com/riscv/riscv-spmp

_Supervisor PMP (sPMP)_ enables supervisor mode to control physical memory access for U mode.

sPMP allows supervisor mode to restrict its own access to memory allocated to lower privilege levels. This can be used to mitigate against privilege escalation attacks, for example.

When combined with H-extension, sPMP can be nested so that the hypervisor can control memory allocations to its guests, and each guest can control its own memory allocations to its workloads.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  
| PMP/ePMP MUST be used to protect M-mode from lower privilege levels.
|===

==== MMU

https://github.com/riscv/riscv-isa-manual/releases/tag/Priv-v1.12[Privileged ISA]

> _Memory management unit (MMU)_ enables memory virtualization for:

Reviewer: ```suggestion
_Memory management unit (MMU)_ enables address translation and protection for:
```

* Isolating an OS from workloads on a system without H-extension (one-stage translation)
* Isolating a hypervisor from a guest, on a system with H-extension (two-stage translation)
 
[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  

> | PMP/ePMP MUST be used to protect M-mode from lower privilege levels.

Reviewer: ```suggestion
| Either PMP/ePMP or MTT MUST be used to protect M-mode from lower privilege levels.
```

|===

==== Supervisor domains

https://github.com/riscv/riscv-smmtt

Supervisor domains allow software components on the same hart to be developed, certified, deployed and attested independently of each other.

A supervisor domain is a compartment above M-mode, physically isolated - memory, execution state, and devices - from other supervisor domains regardless of privilege level (below M-mode). Isolation and context switching between supervisor domains are managed by M-mode firmware. 

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NN
| Either PMP/ePMP or MTT MUST be used to enforce physical memory isolation boundaries for supervisor domains, and to protect machine mode from any supervisor domain.

|===

PMP can be used for more static and deterministic use cases.

MTT can be used where more fine grained dynamic resource management across supervisor domain boundaries is required.

[width=100%]
[%header, cols="5,20"]
|===

| ID#     
| Requirement

| CAT_NNN

> | A system supporting supervisor domains MUST support supervisor domain extensions for interrupts and performance counters, and SHOULD support supervisor domain extensions for external debug.

Reviewer: ```suggestion
| A system supporting supervisor domains MUST support supervisor domain extensions for interrupts (Smsdia) and performance counters (TBD), and SHOULD support supervisor domain extensions for external debug (Smsdsd TBD).
```

External debug: https://github.com/riscv-non-isa/riscv-external-debug-security +
Performance counters: 

These extensions enable management of interrupts, external debug, and performance counters across supervisor domain boundaries. M-mode firmware should context switch hart HPM event/counters to manage isolation of performance counters:

* External debug can be enabled for one supervisor domain without affecting other supervisor domains
* M-mode firmware manage interrupt routing and preemption across supervisor domain boundaries
* M-mode firmware can ensure that performance counters cannot be used by software in one supervisor domain to measure operations in other supervisor domains 

==== MTT

https://github.com/riscv/riscv-smmtt

The _memory tracking table (MTT)_ is a memory structure managed by machine mode, tracking memory ownership across supervisor domains. It is designed to enable fine grained dynamic memory management across supervisor domain boundaries, with policy typically set by a hypervisor in a hosting domain responsible for resource management.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN

> | PMP/ePMP MUST be used to protect M-mode from lower privilege levels

Reviewer: ```suggestion
| Either PMP/ePMP or MTT MUST be used to protect M-mode from lower privilege levels
```

| CAT_NNN  
| MTT configurations MUST only be directly accessible to machine mode
|===

==== IOPMP

https://github.com/riscv-non-isa/iopmp-spec

IOPMP is a system level component providing physical memory access control for device initiated transactions, complementing PMP and sPMP rules. 

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  
| A system which supports PMP/ePMP, or sPMP, MUST implement IOPMP for device access control.

| CAT_NNN
| IOPMP configurations MUST only be directly accessible to machine mode.
|===

==== IOMTT

https://github.com/riscv/riscv-smmtt

IOMTT is a system level component providing physical memory access control for device initiated transactions, complementing MTT rules.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  
| A system which supports MTT MUST implement IOMTT for device initiated access control.

| CAT_NNN
| IOMTT configurations MUST only be directly accessible to machine mode.

| CAT_NNN

> | A system which implements IOMTT MUST also implement IOPMP to protect device initiated accesses to M-mode memory.

Reviewer: ```suggestion
| A system which implements IOMTT MAY also implement IOPMP to protect device initiated accesses to M-mode memory.
```

|===

==== IOMMU

https://github.com/riscv-non-isa/riscv-iommu

IOMMU is a system level component providing virtual memory access for device initiated transactions, complementing MMU translation rules.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN
| Systems supporting MMU SHOULD also support IOMMU

| CAT_NNN  
| Systems supporting IOMMU MUST also enforce physical memory access control for M-mode memory against device initiated transactions (e.g. via IOMTT, IOPMP or equivalent).

|===

=== Software enforced memory tagging

https://github.com/riscv/riscv-j-extension

> _Memory tagging (MT)_, or pointer masking, is a technique which can improve the memory safety of an application. A part of the effective address of a pointer can be masked off, and used as a tag indicating intended ownership or state of a pointer. The tag can be used to track accesses across different regions, as well as protecting against pointer misuse such as "use after free".

Reviewer: ```suggestion
_Memory tagging (MT)_, is a technique which can improve the memory safety of an application. A part of the effective address of a pointer can be masked off, and used as a tag indicating intended ownership or state of a pointer. The tag can be used to track accesses across different regions, as well as protecting against pointer misuse such as "use after free". The pointer masking should use the proposed J-extension pointer masking extension (Zjpm).
```

With software based memory tagging the access rules encoded in tags are enforced by software (compiler).

See also hardware enforced memory tagging below.

=== Control flow integrity

https://github.com/riscv/riscv-cfi

> Control-flow Integrity (CFI) capabilities help defend against Return-Oriented Programming (ROP) and Call/Jump-Oriented Programming (COP/JOP) style control-flow subversion attacks, where an attacker attempts to modify return addresses or call addresses to redirect a victim to code controlled by the attacker.

Reviewer: ```suggestion
Control-flow Integrity (CFI) capabilities help defend against Return-Oriented Programming (ROP) and Call/Jump-Oriented Programming (COP/JOP) style control-flow subversion attacks, where an attacker attempts to modify return addresses or call/jump address to redirect a victim to code reused by the attacker.

These attack methodologies use code sequences in authorized modules, with at least one instruction in the sequence being a control transfer instruction that depends on attacker-controlled data either in the return stack or in memory used to obtain the target address for a call or jump. Attackers stitch these sequences together by diverting the control flow instructions (e.g., JALR, C.JR, C.JALR), from their original target address to a new target via modification in the return stack or in the memory used to obtain the jump/call target address.
```

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NN
| Either PMP/ePMP or MTT MUST be used to enforce physical memory isolation boundaries for supervisor domains, and to protect machine mode from any supervisor domain.

|===

PMP can be used for more static and deterministic use cases.

MTT can be used where more fine grained dynamic resource management across supervisor domain boundaries is required.

[width=100%]
[%header, cols="5,20"]
|===

| ID#     
| Requirement

| CAT_NNN
| A system supporting supervisor domains MUST support supervisor domain extensions for interrupts (Smsdia) and performance counters (TBD), and SHOULD support supervisor domain extensions for external debug (Smsdsd TBD).

|===

Interrupts: https://github.com/riscv/riscv-smmtt +
External debug: https://github.com/riscv-non-isa/riscv-external-debug-security +
Performance counters: 

These extensions enable management of interrupts, external debug, and performance counters across supervisor domain boundaries. M-mode firmware should context switch hart HPM event/counters to manage isolation of performance counters:

> 

Reviewer 2: Since they are context switched, there is no ISA extension required for performance counters.

* M-mode firmware can ensure that performance counters cannot be used by software in one supervisor domain to measure operations in other supervisor domains 

==== MTT

https://github.com/riscv/riscv-smmtt

The _memory tracking table (MTT)_ is a memory structure managed by machine mode, tracking memory ownership across supervisor domains. It is designed to enable fine grained dynamic memory management across supervisor domain boundaries, with policy typically set by a hypervisor in a hosting domain responsible for resource management.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN
| Either PMP/ePMP or MTT MUST be used to protect M-mode from lower privilege levels

| CAT_NNN  
| MTT configurations MUST only be directly accessible to machine mode
|===

==== IOPMP

https://github.com/riscv-non-isa/iopmp-spec

IOPMP is a system level component providing physical memory access control for device initiated transactions, complementing PMP and sPMP rules. 

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  
| A system which supports PMP/ePMP, or sPMP, MUST implement IOPMP for device access control.

> 

Reviewer 2: Would it be better stated as : "A system MUST implement either MTT and/or IOPMP to control device initiated memory accesses"

| CAT_NNN
| IOPMP configurations MUST only be directly accessible to machine mode.
|===

==== IOMTT

https://github.com/riscv/riscv-smmtt

IOMTT is a system level component providing physical memory access control for device initiated transactions, complementing MTT rules.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  

> | A system which supports MTT MUST implement IOMTT for device initiated access control.

Reviewer 2: MAY?

> | A system which supports MTT MUST implement IOMTT for device initiated access control.

Reviewer 2: "device initiated access control" -> "access control for device initiated memory accesses"

| CAT_NNN
| A system which implements IOMTT MAY also implement IOPMP to protect device initiated accesses to M-mode memory.

|===

==== IOMMU

https://github.com/riscv-non-isa/riscv-iommu

IOMMU is a system level component providing virtual memory access for device initiated transactions, complementing MMU translation rules.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN
| Systems supporting MMU SHOULD also support IOMMU

| CAT_NNN  
| Systems supporting IOMMU MUST also enforce physical memory access control for M-mode memory against device initiated transactions (e.g. via IOMTT, IOPMP or equivalent).

|===

=== Software enforced memory tagging

https://github.com/riscv/riscv-j-extension

_Memory tagging (MT)_, is a technique which can improve the memory safety of an application. A part of the effective address of a pointer can be masked off, and used as a tag indicating intended ownership or state of a pointer. The tag can be used to track accesses across different regions, as well as protecting against pointer misuse such as "use after free". The pointer masking should use the proposed J-extension pointer masking extension (Zjpm).

> 

Reviewer 2: Zjpm is not an extension name.

[[chapter3]]

== RISC-V security building blocks

> This chapter outlines brief descriptions of Risc-V security building blocks discussed in this specification, together with general guidelines and links to technical specifications.

Reviewer 3: ```suggestion
This chapter outlines brief descriptions of RISC-V security building blocks discussed in this specification, together with general guidelines and links to technical specifications.
```

> See also the reference use cases chapter of this specification for common examples of how Risc-V security building blocks can be combined. 

Reviewer: ```suggestion
See also the reference use cases chapter of this specification for common examples of how RISC-V security building blocks can be combined. 
```

=== Isolation

> Isolation enable separation of software components executing on a Hart, as well as device assignment. Risc-V enables:

Reviewer: ```suggestion
Isolation enable separation of software components executing on a Hart, as well as device assignment. RISC-V enables:
```

> 

Reviewer: addressed

* Privilege based isolation
* Physical memory access control (hart and device initiated accesses)
* Virtual memory management (hart and device virtualization)
* Hypervisor extension 
* Supervisor domains 

==== Privilege levels

https://github.com/riscv/riscv-isa-manual/releases/tag/Priv-v1.12[Privileged ISA]

Standard privilege levels - Machine mode (M), Supervisor mode (S), and User mode (U) - enable separation of more privileged software from less privileged software.

==== Hypervisor extension

https://github.com/riscv/riscv-isa-manual/releases/tag/Priv-v1.12[Privileged ISA]

> Hypervisor extension supports standard supervisor level hypervisors. It extends S mode into Hypervisor supervisor mode (HS), and a virtual supervisor mode (VS) for guests. It also extends U mode into standard user mode (U) and virtual user mode (VU).

Reviewer 2: Hypervisor supervisor mode -> Hypervisor mode or Hypervisor-extended supervisor mode.

> Hypervisor extension supports standard supervisor level hypervisors. It extends S mode into Hypervisor supervisor mode (HS), and a virtual supervisor mode (VS) for guests. It also extends U mode into standard user mode (U) and virtual user mode (VU).

Reviewer: ```suggestion
Hypervisor extension supports standard supervisor level hypervisors. It extends S mode into Hypervisor-extended supervisor mode (HS), and a virtual supervisor mode (VS) for guests. It also extends U mode into standard user mode (U) and virtual user mode (VU).
```

> Isolation of guests is typically enforced using MMU and virtual memory management.  Memory translation is in effect in VS and VU modes, but not in standard U mode. 

Reviewer 3: By memory translation, do you mean second stage page table based translation?

Reviewer 2: "Isolation of guests is ..." -> "Isolation of guests is enforced using two-stage address translation and protection."

"Memory translation is in effect in VS and VU, but not..." -> "Two-stage address translation and protection is in effect in VS and VU modes".

> Isolation of guests is typically enforced using MMU and virtual memory management.  Memory translation is in effect in VS and VU modes, but not in standard U mode. 

Reviewer: ```suggestion
Isolation of guests is enforced using two-stage address translation and protection.  Two-stage address translation and protection is in effect in VS and VU modes. 
```

> Alternatively sPMP can be used instead of MMU to support static partition hypervisors, for example on systems with hard and deterministic real time requirements. 

Reviewer: ```suggestion
Alternatively sPMP can be used instead of MMU to support static partition hypervisors, for example on systems with hard and deterministic real time requirements [Note -The sPMP for Hypervisor extension has not been specified yet.] . 
```

> 

Reviewer 2: The spmp for hypervisor extension has not been specified yet.

|===

Individual access controlled regions can be locked until the next system reset to create temporal isolation boundaries, such as protecting immutable boot code. 

ePMP extends PMP protection by allowing machine mode to restrict its own access to memory allocated to lower privilege levels. This can be used to mitigate against privilege escalation attacks, for example.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  
| If PMP is supported then ePMP MUST be supported.
|===

==== sPMP

https://github.com/riscv/riscv-spmp

_Supervisor PMP (sPMP)_ enables supervisor mode to control physical memory access for U mode.

sPMP allows supervisor mode to restrict its own access to memory allocated to lower privilege levels. This can be used to mitigate against privilege escalation attacks, for example.

When combined with H-extension, sPMP can be nested so that the hypervisor can control memory allocations to its guests, and each guest can control its own memory allocations to its workloads.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  
| PMP/ePMP MUST be used to protect M-mode from lower privilege levels.
|===

> 

Reviewer: updated to say sPMP, 1st stage or g-stage

==== MMU

https://github.com/riscv/riscv-isa-manual/releases/tag/Priv-v1.12[Privileged ISA]

_Memory management unit (MMU)_ enables memory virtualization for:

> 

Reviewer 2: virtualization -> address translation and protection

Reviewer: addressed

* Isolating an OS from workloads on a system without H-extension (one-stage translation)

> * Isolating a hypervisor from a guest, and a guest from workloads, on a system with H-extension (two-stage translation)

Reviewer 3: Within a guest, first stage translation applies, so I would suggest:

```suggestion
* Isolating a hypervisor from a guest, on a system with H-extension (two-stage translation)
```

 
[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  
| PMP/ePMP MUST be used to protect M-mode from lower privilege levels.
|===

> 

Reviewer: we can clean this up - for now updated it to cover both PMP and MTT options.

> | PMP/ePMP MUST be used to protect M-mode from lower privilege levels.

Reviewer 3: I believe you meant to say that the requirement is that the MMU must be used?

==== Supervisor domains

https://github.com/riscv/riscv-smmtt

> Supervisor domains allow software components on the same hart to be developed, certified, deployed and attested independently of each other.

Reviewer 3: I'm pretty sure you're going to get a lot of bike shedding with the `certified` part here ;-)

> A supervisor domain is a compartment above M-mode, physically isolated - memory, execution state, and devices - from other supervisor domains regardless of privilege level (above M-mode). Isolation and context switching between supervisor domains are managed by M-mode firmware. 

Reviewer: ```suggestion
A supervisor domain is a compartment above M-mode, physically isolated - memory, execution state, and devices - from other supervisor domains regardless of privilege level (below M-mode). Isolation and context switching between supervisor domains are managed by M-mode firmware. 
```

> 

Reviewer 2: above M-mode -> below M-mode.

Reviewer: addressed

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  
| PMP/ePMP MUST be used to protect machine mode from any supervisor domain.

| CAT_NN
| Either PMP/ePMP or MTT MUST be used to enforce physical memory isolation boundaries for supervisor domains.

> 

Reviewer 2: The two should be merged. The second is stronger - either PMP/ePMP or MTT

Reviewer: addressed

|===

PMP can be used for more static and deterministic use cases.

MTT can be used where more fine grained dynamic resource management across supervisor domain boundaries is required.

[width=100%]
[%header, cols="5,20"]
|===

| ID#     
| Requirement

| CAT_NNN
| A system supporting supervisor domains MUST support supervisor domain extensions for interrupts and performance counters, and SHOULD support supervisor domain extensions for external debug.

> 

Reviewer 2: Adding named extensions would help.

Reviewer: put names or placeholders

|===

Interrupts: https://github.com/riscv/riscv-smmtt +
External debug: https://github.com/riscv-non-isa/riscv-external-debug-security +
Performance counters: 

These extensions enable management of interrupts, external debug, and performance counters across supervisor domain boundaries:

* External debug can be enabled for one supervisor domain without affecting other supervisor domains
* M-mode firmware manage interrupt routing and preemption across supervisor domain boundaries

> * M-mode firmware can ensure that performance counters cannot be used by software in one supervisor domain to measure operations in other supervisor domains 

Reviewer 2: Performance counters should not need a new extension. Maybe state, that the M-mode firmware should context switch hart HPM event/counters to acheive this goal.

Reviewer: added a note.

==== MTT

https://github.com/riscv/riscv-smmtt

The _memory tracking table (MTT)_ is a memory structure managed by machine mode, tracking memory ownership across supervisor domains. It is designed to enable fine grained dynamic memory management across supervisor domain boundaries, with policy typically set by a hypervisor in a hosting domain responsible for resource management.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN
| PMP/ePMP MUST be used to protect M-mode from lower privilege levels

> 

Reviewer 2: PMP/ePMP or MTT

| CAT_NNN  
| MTT configurations MUST only be directly accessible to machine mode
|===

==== IOPMP

https://github.com/riscv-non-isa/iopmp-spec

IOPMP is a system level component providing physical memory access control for device initiated transactions, complementing PMP and sPMP rules. 

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  
| A system which supports PMP/ePMP, or sPMP, MUST implement IOPMP for device access control.

| CAT_NNN
| IOPMP configurations MUST only be directly accessible to machine mode.
|===

==== IOMTT

https://github.com/riscv/riscv-smmtt

IOMTT is a system level component providing physical memory access control for device initiated transactions, complementing MTT rules.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  

> | A system which supports MTT MUST implement IOMTT for device initiated access control.

Reviewer 3: Can IOMTT be implemented without an IOMMU? I think both are required.

Reviewer 2: IOMMU is not required for IOMTT. The IOMMU may be disabled or not-implemented. The IOMTT are required to protect against implicit accesses from IOMMU itself as the IOMMU are controlled by the supervisor domains - similar to how MMU is controlled by the SD.

Reviewer: per discussion above - leaving requirement as written

| CAT_NNN
| IOMTT configurations MUST only be directly accessible to machine mode.

| CAT_NNN
| A system which implements IOMTT MUST also implement IOPMP to protect device initiated accesses to M-mode memory.

> 

Reviewer 2: Systems that implements IOMTT MUST -> MAY

Reviewer: addressed

|===

==== IOMMU

https://github.com/riscv-non-isa/riscv-iommu

IOMMU is a system level component providing virtual memory access for device initiated transactions, complementing MMU translation rules.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN
| Systems supporting MMU SHOULD also support IOMMU

| CAT_NNN  

> | Systems supporting IOMMU MUST also support IOPMP, an MAY support IOMTT, to enforce physical memory access control for device initiated transactions.

Reviewer: ```suggestion
| Systems supporting IOMMU MUST also enforce physical memory access control for M-mode memory against device initiated transactions (e.g. via IOMTT, IOPMP or equivalent).
```

> | Systems supporting IOMMU MUST also support IOPMP, an MAY support IOMTT, to enforce physical memory access control for device initiated transactions.

Reviewer 3: I would argue that a system that support IOMMU *and* SMMTT should not require IOPMP.

Reviewer 2: IOPMP or IOMTT suffices. Both are not required.

|===

=== Software enforced memory tagging

https://github.com/riscv/riscv-j-extension

_Memory tagging (MT)_, or pointer masking, is a technique which can improve the memory safety of an application. A part of the effective address of a pointer can be masked off, and used as a tag indicating intended ownership or state of a pointer. The tag can be used to track accesses across different regions, as well as protecting against pointer misuse such as "use after free".

> 

Reviewer 2: Maybe we should separate memory tagging and pointer masking.
Pointer masking uses "non implemented bits in the effective address" - like TBI - so this test needs updating.

Reviewer: addressed

With software based memory tagging the access rules encoded in tags are enforced by software (compiler).

See also hardware enforced memory tagging below.

=== Control flow integrity

https://github.com/riscv/riscv-cfi

Control-flow Integrity (CFI) capabilities help defend against Return-Oriented Programming (ROP) and Call/Jump-Oriented Programming (COP/JOP) style control-flow subversion attacks, where an attacker attempts to modify return addresses or call addresses to redirect a victim to code controlled by the attacker.

> 

Reviewer 2: call address -> call/jump address. 
"code controlled by attacker" is not right since ROP/JOP are used when attacker cannot inject code - its a code-reuse attack.

These attack methodologies use code sequences in authorized modules, with at least one instruction in the sequence being a control transfer instruction that depends on attacker-controlled data either in the return stack or in memory used to obtain the target address for a call or jump. Attackers stitch these sequences together by diverting the control flow instructions (e.g., `JALR`, `C.JR`, `C.JALR`), from their original target address to a new target via modification in the return stack or in the memory used to obtain the jump/call target address.

Reviewer: addressed

Risc-V provides two defenses:

> * Shadow stacks - protect return addresses on call stacks

Reviewer 2: Zicfiss
